packages/alchemy/src/middleware/gas-fees.ts

```diff
@@ -1,33 +1,44 @@
+import { applyFeeOption, type BigNumberish } from "@alchemy/aa-core";
 import type { AlchemyProvider } from "../provider.js";
 import type { ClientWithAlchemyMethods } from "./client.js";
 
-export const withAlchemyGasFeeEstimator = <P extends AlchemyProvider>(
-  provider: P,
-  baseFeeBufferPercent: bigint,
-  maxPriorityFeeBufferPercent: bigint
-): P => {
-  provider.withFeeDataGetter(async () => {
-    const block = await provider.rpcClient.getBlock({ blockTag: "latest" });
-    const baseFeePerGas = block.baseFeePerGas;
-    if (baseFeePerGas == null) {
-      throw new Error("baseFeePerGas is null");
-    }
-    const priorityFeePerGas = BigInt(
-      // it's a fair assumption that if someone is using this Alchemy Middleware, then they are using Alchemy RPC
-      await (provider.rpcClient as ClientWithAlchemyMethods).request({
-        method: "rundler_maxPriorityFeePerGas",
-        params: [],
-      })
-    );
+export const withAlchemyGasFeeEstimator = (
+  provider: AlchemyProvider
+): AlchemyProvider => {
+  provider.withFeeDataGetter(async (struct, overrides, feeOptions) => {
+    const maxPriorityFeePerGas =
+      overrides?.maxPriorityFeePerGas != null
+        ? overrides?.maxPriorityFeePerGas
+        : // it's a fair assumption that if someone is using this Alchemy Middleware, then they are using Alchemy RPC
+          applyFeeOption(
+            await (provider.rpcClient as ClientWithAlchemyMethods).request({
+              method: "rundler_maxPriorityFeePerGas",
+              params: [],
+            }),
+            feeOptions?.maxPriorityFeePerGas
+          );
```

Comment:
seeing this confuses me a bit (and I imagine it would be confusing to a new dev). 

why do both `overrides` and `feeOptions` need to have a value for something like `maxPriorityFeePerGas`? how would I know without looking at the code that the value in `overrides` takes priority over the one in `feeOptions`?

Reply:
we should indicate on the documentation but overrides is a one-time overrides, while feeOptions is at a class level globally. So makes sense for overrides to be override the feeoptions on the provider.

Reply:
okay with that in mind, I actually recommend:
1. having a boolean like `applyProviderFeeOptions` instead of passing around the object itself
2. make the provider's `feeOptions` field readonly instead of protected so it's accessible

where I get confused a dev is why the class-level fields can't just be read from the provider directly if we have its context in any of these middleware decorators. passing them in again and again kind of defeats the point of having them listed as a class field in the first place.

The boolean captures the necessary intent of the developer more succinctly and clearly IMO. wdyt? cc: @moldy530